common/proxy/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use ipnet::IpNet;
+use std::{env, net::IpAddr};
+
+/// Represents a possible matching entry for an IP address
+#[derive(Clone, Debug)]
+enum Ip {
+    Address(IpAddr),
+    Network(IpNet),
+}
+
+/// A wrapper around a list of IP cidr blocks or addresses with a [IpMatcher::contains] method for
+/// checking if an IP address is contained within the matcher
+#[derive(Clone, Debug, Default)]
+struct IpMatcher(Vec<Ip>);
+
+/// A wrapper around a list of domains with a [DomainMatcher::contains] method for checking if a
+/// domain is contained within the matcher
+#[derive(Clone, Debug, Default)]
+struct DomainMatcher(Vec<String>);
+
+/// A configuration for filtering out requests that shouldn't be proxied
+#[derive(Clone, Debug, Default)]
+struct NoProxy {
+    ips: IpMatcher,
+    domains: DomainMatcher,
+}
+
+pub struct Proxy {
+    http_proxy: Option<String>,
+    https_proxy: Option<String>,
+    no_proxy: Option<NoProxy>,
+}
+
+impl Proxy {
+    pub fn new() -> Self {
+        let http_proxy = env::var("http_proxy")
+            .or_else(|_| env::var("HTTP_PROXY"))
+            .ok();
+        let https_proxy = env::var("https_proxy")
+            .or_else(|_| env::var("HTTPS_PROXY"))
+            .ok();
+        let no_proxy = NoProxy::new();
+
+        Self {
+            http_proxy,
+            https_proxy,
+            no_proxy,
+        }
+    }
+
+    pub fn http(&self, host: &str) -> Option<&str> {
+        if let Some(no_proxy) = &self.no_proxy {
+            if no_proxy.contains(host) {
+                return None;
+            }
+        }
+        self.http_proxy.as_deref()
+    }
+
+    pub fn https(&self, host: &str) -> Option<&str> {
+        if let Some(no_proxy) = &self.no_proxy {
+            if no_proxy.contains(host) {
+                return None;
+            }
+        }
+        self.https_proxy.as_deref()
+    }
+}
+
+impl NoProxy {
+    /// Returns a new no proxy configration if the no_proxy/NO_PROXY environment variable is set.
+    /// Returns None otherwise
+    fn new() -> Option<Self> {
+        let raw = env::var("no_proxy")
+            .or_else(|_| env::var("NO_PROXY"))
+            .unwrap_or_default();
+        if raw.is_empty() {
+            return None;
+        }
+        let mut ips = Vec::new();
+        let mut domains = Vec::new();
+        let parts = raw.split(',');
+        for part in parts {
+            match part.parse::<IpNet>() {
```

Comment:
should we use a more common dependency such as Url? is this adding a dependency?

Reply:
this looks like it's parsing IP CIDR blocks, so I don't think Url helps here : (

Reply:
My point is, we don't really need to parse anything, we can get the Host and see if Rust can parse it into an IPAddress or not.

Reply:
As Philip mentioned this is specifically so that we can parse and handle ip cidr blocks so its needed in order to properly handle the no_proxy spec

Reply:
NO_PROXY can be an IP CIDR block, not just an IP addr. like `NO_PROXY=10.15.2.0/16`. trying to parse "10.15.2.0/16" as an IP addr fails: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1d22e3182355c3da31fb86357974263f

Reply:
I missed line 89 and focused only on line 91 time to end the day :P